core/src/main/java/com/linecorp/armeria/common/stream/FuseableStreamMessage.java

```diff
@@ -0,0 +1,280 @@
+    public void subscribe(Subscriber<? super U> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        source.subscribe(new FuseableSubscriber<U>(subscriber, function), executor, options);
+    }
+
+    @Override
+    public void abort() {
+        source.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        source.abort(cause);
+    }
+
+    private static final class FuseableSubscriber<U> implements Subscriber<Object>, Subscription {
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<FuseableSubscriber, Subscription> upstreamUpdater =
+                AtomicReferenceFieldUpdater
+                        .newUpdater(FuseableSubscriber.class, Subscription.class, "upstream");
+
+        private final Subscriber<? super U> downstream;
+        private final MapperFunction<Object, U> function;
+
+        @Nullable
+        private volatile Subscription upstream;
+
+        FuseableSubscriber(Subscriber<? super U> downstream, MapperFunction<Object, U> function) {
+            requireNonNull(downstream, "downstream");
+            this.downstream = downstream;
+            this.function = function;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (!upstreamUpdater.compareAndSet(this, null, subscription)) {
+                subscription.cancel();
+                throw new IllegalStateException("Subscription already set!");
+            }
+            upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(Object item) {
+            requireNonNull(item, "item");
+
+            final Subscription upstream = this.upstream;
+            if (upstream == null) {
+                return;
+            }
+            U result = null;
+            try {
+                result = function.apply(item);
+                if (result != null) {
+                    downstream.onNext(result);
+                } else {
+                    StreamMessageUtil.closeOrAbort(item);
+                    upstream.request(1);
+                }
+            } catch (Throwable ex) {
+                StreamMessageUtil.closeOrAbort(item);
+                if (result != null && item != result) {
+                    StreamMessageUtil.closeOrAbort(result);
+                }
+                upstream.cancel();
+                onError(ex);
+            }
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            downstream.onError(throwable);
+        }
+
+        @Override
+        public void onComplete() {
+            downstream.onComplete();
```

Comment:
Aren't there chances that `onComplete` is called before the `downStream` is set?

Reply:
Oops, please forget about this comment. 😅

Reply:
The `downstream` is set in the constructor of `FuseableSubscriber`, so it is a non-null and final value.